components/wallet/lnc.js

```diff
@@ -0,0 +1,190 @@
+import bip39Words from '@/lib/bip39-words'
+import LNC from '@lightninglabs/lnc-web'
+import { Mutex } from 'async-mutex'
+import { string, array, object } from 'yup'
+import { Form, PasswordInput, SubmitButton } from '@/components/form'
+import CancelButton from '@/components/cancel-button'
+import { InvoiceCanceledError, InvoiceExpiredError } from '@/components/payment'
+import { bolt11Tags } from '@/lib/bolt11'
+import { Status } from '@/components/wallet'
+
+export const name = 'lnc'
+
+export const fields = [
+  {
+    name: 'pairingPhrase',
+    label: 'pairing phrase',
+    type: 'password',
+    help: 'We only need permissions for the uri `/lnrpc.Lightning/SendPaymentSync`\n\nCreate a budgeted account with narrow permissions:\n\n```$ litcli accounts create --balance <budget>```\n\n```$ litcli sessions add --type custom --label <your label> --account_id <account_id> --uri /lnrpc.Lightning/SendPaymentSync```\n\nGrab the `pairing_secret_mnemonic` from the output and paste it here.'
+  },
+  {
+    name: 'password',
+    label: 'password',
+    type: 'password',
+    hint: 'encrypts your pairing phrase when stored locally',
+    optional: true
+  }
+]
+
+export const card = {
+  title: 'LNC',
+  subtitle: 'use Lightning Node Connect for LND payments',
+  badges: ['send only', 'non-custodialish', 'budgetable']
+}
+
+const XXX_DEFAULT_PASSWORD = 'password'
+
+export async function validate ({ me, logger, pairingPhrase, password }) {
+  const lnc = await getLNC({ me })
+  try {
+    lnc.credentials.pairingPhrase = pairingPhrase
+    logger.info('connecting ...')
+    // FIXME: this fails with this error:
+    //   Cannot assign to read only property 'undefined' of object '#<Window>'
+    await lnc.connect()
+    logger.ok('connected')
+    logger.info('validating permissions ...')
+    await validateNarrowPerms(lnc)
+    logger.ok('permissions ok')
+    lnc.credentials.password = password || XXX_DEFAULT_PASSWORD
+    logger.info('getting lightning info ...')
+    await lnc.lightning.getInfo()
+    logger.ok('info received')
+  } finally {
+    // FIXME: this fails with this error:
+    //   Cannot read properties of undefined (reading 'wasmClientDisconnect')
+    // uncommented because it shadows the error from lnc.connect()
+    // lnc.disconnect()
+  }
+}
+
+export const schema = object({
+  pairingPhrase: array()
+    .transform(function (value, originalValue) {
+      if (this.isType(value) && value !== null) {
+        return value
+      }
+      return originalValue ? originalValue.split(/[\s]+/) : []
+    })
+    .of(string().trim().oneOf(bip39Words, ({ value }) => `'${value}' is not a valid pairing phrase word`))
```

Comment:
This creates a separate error for every invalid word which looks bad

![2024-06-25-155030_725x113_scrot](https://github.com/stackernews/stacker.news/assets/27162016/e41ec2f5-772a-4424-8b61-1565a4c67ecb)

Reply:
Done in 9af8e633